File: ios-app/Plantgotchi/Models/Achievement.swift
import Foundation
import GRDB

/// A user achievement record.
struct Achievement: Codable, Identifiable, Equatable {
    var id: String
    var userId: String
    var achievementKey: String
    var points: Int
    var unlockedAt: String?
    var metadata: String?

    init(
        id: String = UUID().uuidString,
        userId: String,
        achievementKey: String,
        points: Int,
        unlockedAt: String? = nil,
        metadata: String? = nil
    ) {
        self.id = id
        self.userId = userId
        self.achievementKey = achievementKey
        self.points = points
        self.unlockedAt = unlockedAt
        self.metadata = metadata
    }
}

// MARK: - GRDB Records

extension Achievement: TableRecord, FetchableRecord, PersistableRecord {
    static let databaseTableName = "achievements"

    enum Columns: String, ColumnExpression {
        case id
        case userId = "user_id"
        case achievementKey = "achievement_key"
        case points
        case unlockedAt = "unlocked_at"
        case metadata
    }

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case achievementKey = "achievement_key"
        case points
        case unlockedAt = "unlocked_at"
        case metadata
    }
}

// MARK: - Achievement Definitions

/// Known achievements with their point values.
enum AchievementDef: String, CaseIterable {
    case firstSeed     = "first_seed"
    case firstHarvest  = "first_harvest"
    case tenPlants     = "ten_plants"
    case firstTop      = "first_top"
    case firstLST      = "first_lst"
    case speedGrow     = "speed_grow"
    case firstGram     = "first_gram"
    case bigYield100g  = "big_yield_100g"
    case weekStreak    = "week_streak"
    case fiveStrains   = "five_strains"

    var key: String { rawValue }

    var points: Int {
        switch self {
        case .firstSeed:    return 10
        case .firstHarvest: return 50
        case .tenPlants:    return 30
        case .firstTop:     return 20
        case .firstLST:     return 20
        case .speedGrow:    return 100
        case .firstGram:    return 25
        case .bigYield100g: return 75
        case .weekStreak:   return 15
        case .fiveStrains:  return 40
        }
    }

    var label: String {
        switch self {
        case .firstSeed:    return "First Seed"
        case .firstHarvest: return "First Harvest"
        case .tenPlants:    return "Ten Plants"
        case .firstTop:     return "First Top"
        case .firstLST:     return "First LST"
        case .speedGrow:    return "Speed Grow"
        case .firstGram:    return "First Gram"
        case .bigYield100g: return "Big Yield (100g)"
        case .weekStreak:   return "Week Streak"
        case .fiveStrains:  return "Five Strains"
        }
    }
}

File: android-app/app/src/main/java/com/plantgotchi/app/model/Achievement.kt
package com.plantgotchi.app.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "achievements",
    indices = [Index(value = ["user_id", "achievement_key"], unique = true)]
)
data class Achievement(
    @PrimaryKey
    val id: String,

    @ColumnInfo(name = "user_id")
    val userId: String,

    @ColumnInfo(name = "achievement_key")
    val achievementKey: String,

    val points: Int = 0,

    @ColumnInfo(name = "unlocked_at")
    val unlockedAt: String? = null,

    val metadata: String? = null, // JSON string
)

data class AchievementDef(
    val key: String,
    val name: String,
    val points: Int,
    val description: String,
)

val ACHIEVEMENT_DEFS = listOf(
    AchievementDef("firstSeed", "First Seed", 10, "Start your first plant"),
    AchievementDef("firstHarvest", "First Harvest", 50, "Complete your first harvest"),
    AchievementDef("tenPlants", "Green Thumb", 30, "Grow 10 plants"),
    AchievementDef("firstTop", "First Top", 20, "Top a plant for the first time"),
    AchievementDef("firstLST", "First LST", 20, "Apply LST for the first time"),
    AchievementDef("speedGrow", "Speed Grower", 100, "Complete a grow in record time"),
    AchievementDef("firstGram", "First Gram", 25, "Harvest your first gram"),
    AchievementDef("bigYield100g", "Big Yield", 75, "Harvest 100g or more from a single plant"),
    AchievementDef("weekStreak", "Week Streak", 15, "Log activity 7 days in a row"),
    AchievementDef("fiveStrains", "Strain Collector", 40, "Grow 5 different strains"),
)
